Fix React library visiting logic

calebmer opened the pull request:
> I’m learning the Prepack/React Fusion codebase, so incorrect assumptions and limited a understanding of systems on my part are both very likely.

While I was experimenting with the React Compiler codebase I found the following invariant violation with the test case after that when using `create-element` as the `reactOutput`.

```
Invariant Violation: value must have been visited
```

```js
const React = require("react");

__evaluatePureFunction(() => {
  function Foo(props) {
    return <React.Fragment />;
  }

  __optimizeReactComponentTree(Foo);

  module.exports = Foo;
});
```

[[Prepack REPL]](https://prepack.io/repl.html#MYewdgzgLgBASgUwIbFgXhgJwQRwK4CW2AFAETYpSkCUA3AFD0D6TCAbkgDZ5JQIAKebADE8YVAXDFi1GGgB8MAN70YMAGZiJ4GMJAhiAB0whDEWSrVrsUIWBgAeRJQB0wzEgDmAWwRhYAPTyDGoAvoxqLKZQBN4EAF4IzqgAwiDehuB+UAAq2AjEeiB0ETDeIAAmeJwILggAHpmYUBByuvoMoSVAA)

After digging in more, I discovered that this was the same issue which required some clowniness in the React Compiler tests to workaround.

https://github.com/facebook/prepack/blob/e3f4262c9dafce766835acdaff39af04de231fae/test/react/setupReactTests.js#L116-L119

The problem with my example is that the React library was not being visited when a `React.Fragment` appeared even though the serializer needed the library to be visited to output code.

This diff makes sure we visit the React library when we see a `React.Fragment` and makes sure we don’t over visit the React library in `jsx` mode.

I also refactored this area of the code and added comments. Let me know if my refactors or comments don’t make sense.

calebmer commented on src/serializer/ResidualHeapVisitor.js:
```diff
@@ -1223,21 +1223,6 @@ export class ResidualHeapVisitor {
     for (let moduleValue of this.modules.initializedModules.values()) this.visitValue(moduleValue);
 
-    if (this.realm.react.enabled) {
-      let fixpoint_rerun = () => {
```
Why were we using this `fixpoint_rerun` recursion? I moved the visiting logic into `ResidualReactElementVisitor` along with the other visiting logic for that class and didn’t observe any test failures.

trueadm replied:
We may have been doing this in the case where React is behind something that is behind a value that is delayed. If you run your PR on our internal bundle that I sent you the other night, do we hit any issues when compiling it?

calebmer replied:
Internal bundle builds 👍

calebmer commented on src/serializer/ResidualHeapVisitor.js:
```diff
@@ -1333,36 +1318,4 @@ export class ResidualHeapVisitor {
-      }
-    } else if (this.realm.react.output === "create-element") {
-      let logError = () => {
-        this.logger.logError(
```
This error wouldn’t be recoverable since we would get an `invariant` exception in our serializer. So I updated the code to throw a `FatalError` instead.

calebmer commented on src/serializer/ResidualReactElementVisitor.js:
```diff
@@ -111,9 +111,29 @@ export class ResidualReactElementVisitor {
+        // our hoisted React elements. So we need to ensure that we visit our value in a scope above our own to allow
+        // the function to be used in our optimized function prelude. We use our global scope to accomplish this. We are
+        // a "friend" class of `ResidualHeapVisitor` so we call one of its private methods.
+        this.residualHeapVisitor._visitInUnrelatedScope(this.residualHeapVisitor.globalGenerator, reactCreateElement);
```
Visiting this value in the global generator is a bit clowny (read comment), but it is the same thing that the previous `fixpoint_rerun` code did.

Is this be a valid call to `_withScope`? Why? There’s a scary comment on `_withScope` I don’t yet understand which makes me less confident about this.

https://github.com/facebook/prepack/blob/e3f4262c9dafce766835acdaff39af04de231fae/src/serializer/ResidualHeapVisitor.js#L243-L245

NTillmann commented on src/serializer/ResidualReactElementVisitor.js:
```diff
@@ -154,4 +172,13 @@ export class ResidualReactElementVisitor {
+    const reactLibraryObject = this.realm.fbLibraries.react;
+    // if there is no React library, then we should throw and error
+    if (reactLibraryObject === undefined) {
+      throw new FatalError("unable to find React library reference in scope");
```
Never just throw a FatalError without first issuing a CompilerDiagnostics. Doing so should trigger another invariant.

NTillmann replied:
If this indicates an internal error, stick to an invariant.

trueadm replied:
Let's go back to the invariant here, it's an internal Prepack issue if it fails.

calebmer replied:
I had assumed that this error might be thrown when the user did not `require('react')`. Guess we wouldn’t make it to this point if that were the case.

gaearon commented:
Lol this tripped me hard a few days ago. Thanks for fixing.

calebmer commented:
I built a simple fuzzer to help me learn the React Compiler and what its limits are. Running my change through that fuzzer caught a bug in this program:

```js
const React = require("react");

__evaluatePureFunction(() => {
  function Root(props) {
    return (
      <React.Fragment>
        <React.Fragment>
          <div />
        </React.Fragment>
      </React.Fragment>
    );
  }

  __optimizeReactComponentTree(Root);

  module.exports = Root;
});
```

I fixed that bug and added a test case.
